contracts/exchangeV3/TokenGroupRegistry.sol

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: MIT
+pragma solidity =0.8.10;
+
+import "../auth/AdminAuth.sol";
+
+/// @title TokenGroupRegistry - Keeps track of tokens that are in a same group for diff. fee models
+contract TokenGroupRegistry is AdminAuth {
+    /// @dev 0.25% fee as we divide the amount with this number
+    uint256 public constant STANDARD_FEE_DIVIDER = 400;
+    
+    uint256 public constant STABLE_FEE_DIVIDER = 1000;
+    uint256 public constant MAX_FEE_DIVIDER = 50;
+
+    /// @dev maps token address to a registered group it belongs to
+    mapping(address => uint256) public groupIds;
+
+    /// @dev Array of groups where the index is the grouped id and the value is the fee
+    uint256[] public feesPerGroup;
+
+    enum Groups { NOT_LISTED, BANNED, STABLECOIN, ETH_BASED, BTC_BASED}
+
+    error FeeTooHigh(uint256 fee);
+    error GroupNonExistent(uint256 groupId);
+
+    constructor() {
+        feesPerGroup.push(STANDARD_FEE_DIVIDER); // NOT_LISTED
+        feesPerGroup.push(0);                    // BANNED
+        feesPerGroup.push(STABLE_FEE_DIVIDER);   // STABLECOIN
+        feesPerGroup.push(STABLE_FEE_DIVIDER);   // ETH_BASED
+        feesPerGroup.push(STABLE_FEE_DIVIDER);   // BTC_BASED
+    }
+
+    /// @notice Checks if 2 tokens are in the same group and returns the correct exchange fee for the pair
+    function getFeeForTokens(address _sellToken, address _buyToken) public view returns (uint256) {
+        uint256 firstId = groupIds[_sellToken];
+        uint256 secondId = groupIds[_buyToken];
+
+        // check if in the ban list, can just check the first token as we take fee from it
+        if (firstId == uint8(Groups.BANNED)) {
+            return 0;
+        }
+    
+        if (firstId == secondId) {
+            return feesPerGroup[secondId];
+        }
+
+        return STANDARD_FEE_DIVIDER;
+    }
+
+    /////////////////////////////// ONLY OWNER FUNCTIONS ///////////////////////////////
+
+    /// @notice Adds token to an existing group
+    /// @dev This will overwrite if token is part of a different group
+    /// @dev Groups needs to exist to add to it
+    function addTokenInGroup(address _tokenAddr, uint256 _groupId) public onlyOwner {
+        if (_groupId >= feesPerGroup.length) revert GroupNonExistent(_groupId);
```

Comment:
this needs to be > and not >=
ex. feesFerGroup has length 5, so possible groupIds are 0,1,2,3,4
with >= groupId 5 will pass